### PR TITLE
Fix XenServer nested extra configuration

### PR DESCRIPTION
**Pull request by gpordeus**
### Description

This PR fixes a problem when trying to boot up a XenServer VM with a template through an ISO. 

Steps to reproduce:
- Enable `enable.additional.vm.configuration` and add `HVM-boot-params:order` to `allow.additional.vm.configuration.list.xen`.
- Create a XenServer VM with a template.
- Attach an ISO to it.
- Call `updateVirtualMachine` with parameter `extraconfig=HVM-boot-params:order%3Ddcn`
- Start up the VM.

It will not finish, with the following log message:
```
INFO  [c.c.h.x.r.XenServer650Resource] (DirectAgent-71:[ctx-6c7bfe3d]) (logid:faabbafa) Appending user extra configuration settings to VM
ERROR [o.a.c.h.x.ExtraConfigurationUtility] (DirectAgent-71:[ctx-6c7bfe3d]) (logid:faabbafa) Exception caught while setting VM configuration. exception: null
```

After fixing the exception message, we get:

```
ERROR [o.a.c.h.x.ExtraConfigurationUtility] (DirectAgent-25:ctx-0d31d123) (logid:aa40d07d) Exception caught while setting VM configuration: [You tried to add a key-value pair to a map, but that key is already there.]
```

This PR fixes the exception message and changes the Xen API method called from `addTo` to `set` with `get`, letting Java substitute any duplicates in the HashMap.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
I repeated the steps to reproduce the error and verified the VM with template started up sucessfully, booting the ISO.

### CODE REVIEW

**Review by DaanHoogland — COMMENTED** [review]
CLGTM, but don't you think the replacement code should be called inside the `VirtualMachine.add<>()` methods, @gpordeus ?

**Comment by DaanHoogland** [comment]
> @DaanHoogland Thanks for the review.
> 
> > CLGTM, but don't you think the replacement code should be called inside the `VirtualMachine.add<>()` methods, @gpordeus ?
> 
> Do you mean the original methods (`VM.addToVCPUsParams()`, `VM.addToHVMBootParams()`, etc)? It would be better, but those are part of the xenapi :/

right, I thought these were in a wrapper :( ok, don't bother

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by vishesh92** [comment]
@gpordeus Unit tests are failing during build. Can you check & fix those?

**Comment by gpordeus** [comment]
> @gpordeus Unit tests are failing during build. Can you check & fix those?

@DaanHoogland @vishesh92
It seems the test failed because dl.openvm.eu was down at the time. Could you run them again?

```
[INFO] Installing /home/runner/work/cloudstack/cloudstack/framework/quota/pom.xml to /home/runner/.m2/repository/org/apache/cloudstack/cloud-framework-quota/4.20.0.0-SNAPSHOT/cloud-framework-quota-4.20.0.0-SNAPSHOT.pom
16:29:16.334 [main] ERROR org.apache.cloudstack.direct.download.HttpDirectTemplateDownloader - Cannot reach URL: http://dl.openvm.eu/cloudstack/macchinina/x86_64/macchinina-kvm.qcow2.bz2 due to: dl.openvm.eu
java.net.UnknownHostException: dl.openvm.eu
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:229) ~[?:?]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:?]
	at java.net.Socket.connect(Socket.java:609) ~[?:?]
	at java.net.Socket.connect(Socket.java:558) ~[?:?]
	at java.net.Socket.<init>(Socket.java:454) ~[?:?]
	at java.net.Socket.<init>(Socket.java:306) ~[?:?]
	at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:80) ~[commons-httpclient-3.1.jar:?]
	at org.apache.commons.httpclient.protocol.ControllerThreadSocketFactory$1.doit(ControllerThreadSocketFactory.java:91) ~[commons-httpclient-3.1.jar:?]
	at org.apache.commons.httpclient.protocol.ControllerThreadSocketFactory$SocketTask.run(ControllerThreadSocketFactory.java:158) ~[commons-httpclient-3.1.jar:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.965 s <<< FAILURE! - in org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest
Error:  testCheckUrlMetalink(org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest)  Time elapsed: 10.964 s  <<< FAILURE!
java.lang.AssertionError
	at org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest.testCheckUrlMetalink(MetalinkDirectTemplateDownloaderTest.java:34)
```

**Comment by weizhouapache** [comment]
> > @gpordeus Unit tests are failing during build. Can you check & fix those?
> 
> @DaanHoogland @vishesh92
> It seems the test failed because dl.openvm.eu was down at the time. Could you run them again?
> 
> ```
> [INFO] Installing /home/runner/work/cloudstack/cloudstack/framework/quota/pom.xml to /home/runner/.m2/repository/org/apache/cloudstack/cloud-framework-quota/4.20.0.0-SNAPSHOT/cloud-framework-quota-4.20.0.0-SNAPSHOT.pom
> 16:29:16.334 [main] ERROR org.apache.cloudstack.direct.download.HttpDirectTemplateDownloader - Cannot reach URL: http://dl.openvm.eu/cloudstack/macchinina/x86_64/macchinina-kvm.qcow2.bz2 due to: dl.openvm.eu
> java.net.UnknownHostException: dl.openvm.eu
> 	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:229) ~[?:?]
> 	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[?:?]
> 	at java.net.Socket.connect(Socket.java:609) ~[?:?]
> 	at java.net.Socket.connect(Socket.java:558) ~[?:?]
> 	at java.net.Socket.<init>(Socket.java:454) ~[?:?]
> 	at java.net.Socket.<init>(Socket.java:306) ~[?:?]
> 	at org.apache.commons.httpclient.protocol.DefaultProtocolSocketFactory.createSocket(DefaultProtocolSocketFactory.java:80) ~[commons-httpclient-3.1.jar:?]
> 	at org.apache.commons.httpclient.protocol.ControllerThreadSocketFactory$1.doit(ControllerThreadSocketFactory.java:91) ~[commons-httpclient-3.1.jar:?]
> 	at org.apache.commons.httpclient.protocol.ControllerThreadSocketFactory$SocketTask.run(ControllerThreadSocketFactory.java:158) ~[commons-httpclient-3.1.jar:?]
> 	at java.lang.Thread.run(Thread.java:829) ~[?:?]
> Error:  Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.965 s <<< FAILURE! - in org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest
> Error:  testCheckUrlMetalink(org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest)  Time elapsed: 10.964 s  <<< FAILURE!
> java.lang.AssertionError
> 	at org.apache.cloudstack.direct.download.MetalinkDirectTemplateDownloaderTest.testCheckUrlMetalink(MetalinkDirectTemplateDownloaderTest.java:34)
> ```

The server is not down.
But the domain name can not be resolved by some DNS servers (e.g. 8.8.8.8)
Please try 1.1.1.1

@DaanHoogland @NuxRo 
We probably can move templates to download.cloudstack.org

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java [hunk 1]
```diff
@@ -1972,7 +1972,7 @@ protected void finalizeVmMetaData(final VM vm, final VM.Record vmr, final Connec
         // Add configuration settings VM record for User VM instances before creating VM
         Map<String, String> extraConfig = vmSpec.getExtraConfig();
         if (vmSpec.getType().equals(VirtualMachine.Type.User) && MapUtils.isNotEmpty(extraConfig)) {
-            logger.info("Appending user extra configuration settings to VM");
+            logger.info("Appending user extra configuration settings [{}] to [{}].", extraConfig, vmSpec));
```
[the remark this answers is not in the document]
Comment (gpordeus) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java:1975]:
```suggestion
            logger.info("Appending user extra configuration settings [{}] to [{}].", extraConfig, vmSpec);
```

Oops, an extra parenthesis sneaked in there and I hadn't tested them locally again.  It should be fine, now.

[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/hypervisors/xenserver/src/main/java/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java]: my bad to sugest that change

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 xcpng82

**Comment by DaanHoogland** [comment]
> alma9 xcpng82

@gpordeus the test run with these env specs fails a bit, is that expected? what did you use?

**Comment by gpordeus** [comment]
> @gpordeus the test run with these env specs fails a bit, is that expected? what did you use?

I tested it manually on a local ubuntu lab with xcp-ng-8.2 host. I didn't run the integration tests, but I believe these are related to #8586, since the same tests are acting up (`cluster_drs`, `nonstrict_affinity_group`, `register_userdata` and `vm_deployment_planner`).

**Comment by DaanHoogland** [comment]
> > @gpordeus the test run with these env specs fails a bit, is that expected? what did you use?
> 
> I tested it manually on a local ubuntu lab with xcp-ng-8.2 host. I didn't run the integration tests, but I believe these are related to #8586, since the same tests are acting up (`cluster_drs`, `nonstrict_affinity_group`, `register_userdata` and `vm_deployment_planner`).

@weizhouapache , can you review?
@gpordeus , can you run the integration tests to see?

**Review by weizhouapache — APPROVED** [review]
code lgtm

cc @DaanHoogland

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 xenserver-71

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@vishesh92 is this alright to you?

**Review by vishesh92 — APPROVED** [review]
clgtm. didn't test.

**Pull request merged**